### PR TITLE
CopyFile Improvement

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

There have been customer suggestions to make CopyFile easier to work with. Example can be see at https://github.com/DynamoDS/Dynamo/issues/11393. 

CopyFile node now 
- returns true if the copy action succeeded for user.
- returns null if the copy action failed for some reason, the node should be in warning state for user to check reason.
- Validate the input `destinationPath` for certain edge cases, e.g. File Path node in its initial state, CodeBlockNode with some random string
- Update the tooltip of input `destinationPath` to indicate it should be file path, not including directory path for now (although at somepoint we should support that for user convenience)

A few screenshots for different cases, notice the error message has been updated to `Invalid destination file path used as copy path.`
image

image

image

image

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/Libraries/CoreNodes/FileSystem.cs [hunk 1]
```diff
@@ -98,12 +98,28 @@ public static void DeleteFile(string path)
...
-            file.CopyTo(destinationPath, overwrite);
+            try
+            {
+                if (FileExists(destinationPath) || DirectoryExists(destinationPath))
+                {
+                    file.CopyTo(destinationPath, overwrite);
+                }
+                else
+                {
+                    throw new FileNotFoundException("Invalid destinationPath", destinationPath);
```
Comment (QilongTang) [inline src/Libraries/CoreNodes/FileSystem.cs:114]: Will convert this into a resource string soon.

**Comment by QilongTang** [comment]
By checking the regression, I found there is actual change to behavior. I will update the PR shortly

**Review by zeusongit — APPROVED** [review]

File: src/Libraries/CoreNodes/FileSystem.cs [hunk 2]
```diff
@@ -98,12 +98,28 @@ public static void DeleteFile(string path)
         ///     Copies a file.
         /// </summary>
         /// <param name="file">File object to copy</param>
-        /// <param name="destinationPath">String representation of destination path</param>
+        /// <param name="destinationPath">String representation of destination file path</param>
         /// <param name="overwrite">Toggle to overwrite existing files</param>
-        /// <returns name="void">Node performs a task, doesn’t produce an output </returns>
-        public static void CopyFile(FileInfo file, string destinationPath, bool overwrite = false)
+        /// <returns name="bool">Node performs a task, return true of copy action succeed.</returns>
+        public static bool CopyFile(FileInfo file, string destinationPath, bool overwrite = false)
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/FileSystem.cs:104]: @QilongTang this change likely means that any c# packages which called this function from a ZT node will need to be recompiled.

That probably is a very small number.

Comment (QilongTang) [reply src/Libraries/CoreNodes/FileSystem.cs:104]: Thanks for the reminder! @mjkkirschner

**Pull request merged**